python/ray/util/client/worker.py

```diff
@@ -59,23 +71,68 @@ def __init__(self,
         else:
             self.channel = grpc.insecure_channel(conn_str)
 
+        # Retry the connection until the channel responds to something
+        # looking like a gRPC connection, though it may be a proxy.
+        #
+        # This is a finite state machine following the progression of states
+        # a connecting client should check and retry for.
         conn_attempts = 0
         timeout = INITIAL_TIMEOUT_SEC
+        connection_state = CONN_STATE_CHANNEL
```

Comment:
Why do we need a FSM? Why not reset the state to the initial state on connection failure and retry end-to-end?

If the server is up, the connection should be instant so there isn't any benefit in keeping intermediate states around on failure.